lib/puppet/node/environment.rb

```diff
@@ -82,8 +82,12 @@ def self.remote(name)
   # @param name [Symbol] The environment name
   def initialize(name, modulepath, manifest, config_version)
     @name = name
-    @modulepath = modulepath
-    @manifest = manifest
+    @modulepath = modulepath.map { |p| Puppet::FileSystem.expanded_path_string(p) }
+    if manifest && !manifest.is_a?(Symbol) && Puppet::FileSystem.exist?(manifest)
+      @manifest = Puppet::FileSystem.expanded_path_string(manifest)
+    else
+      @manifest = manifest
+    end
```

Comment:
It makes sense to bake the path that is the environment root in the environment. There are probably many use cases that use that as the root for path construction.
What I am wary of though is that differences bubble up. I can imagine environment having a separate method for "file_system_impl_specific_path", and that it is cached in the environment.
That takes the guesswork out of code in general. Did it get the specific one, should it process it again? Hard to know. Thus, we make every use case of file paths responsible for using the right kind of path.
In the loaders the "smart paths", and when creating a tentative path to test with. Since it has explicit knowledge about the paths, there should be no need to expand the parts relative to the root.